File: slack-poster/src/test/java/dev/number6/slackposter/SlackPosterComponentIntegrationTest.java
package dev.number6.slackposter;

import dev.number6.slackposter.dagger.DaggerTestSlackPosterComponent;
import dev.number6.slackposter.dagger.FakeSlackPosterConfigurationModule;
import dev.number6.slackposter.dagger.RecordingHttpAdaptor;
import dev.number6.slackposter.dagger.TestSlackPosterComponent;
import dev.number6.slackposter.model.ChannelSummaryImageBuilder;
import dev.number6.slackposter.model.Chat;
import dev.number6.slackposter.model.PresentableChannelSummary;
import com.amazonaws.services.lambda.runtime.LambdaLogger;
import com.google.gson.Gson;
import org.junit.jupiter.api.Test;

import static org.assertj.core.api.Assertions.assertThat;

public class SlackPosterComponentIntegrationTest {

    TestSlackPosterComponent testee = DaggerTestSlackPosterComponent.create();
    Gson gson = new Gson();

    @Test
    void postCorrectChatToSlack() {
        ChannelSummaryImage image = ChannelSummaryImageBuilder.finalImage().build();
        SlackService service = testee.handler();
        service.handleNewImage(image, new FakeLambdaLogger());
        RecordingHttpAdaptor http = (RecordingHttpAdaptor) testee.getRecordingHttpAdaptor();

        assertThat(http.getPosts()).hasSize(1);
        var chat = gson.fromJson(http.getPosts().get(FakeSlackPosterConfigurationModule.SLACK_POST_MESSAGE_URL),
                Chat.class);
        assertThat(chat.text).isEqualTo(new PresentableChannelSummary(image).getInitialMessageLine());
        assertThat(chat.channel).isEqualTo(FakeSlackPosterConfigurationModule.POSTING_CHANNEL_ID);
    }

    static class FakeLambdaLogger implements LambdaLogger {

        @Override
        public void log(String message) {
            System.out.println(message);
        }

        @Override
        public void log(byte[] message) {
            throw new UnsupportedOperationException("doin' wut now?");
        }
    }
}

File: slack-poster/src/test/kotlin/dev/number6/slackposter/SlackPosterComponentIntegrationTest.kt
package dev.number6.slackposter

import assertk.assertThat
import assertk.assertions.hasSize
import assertk.assertions.isEqualTo
import com.amazonaws.services.lambda.runtime.LambdaLogger
import com.google.gson.Gson
import dev.number6.slackposter.dagger.DaggerTestSlackPosterComponent
import dev.number6.slackposter.dagger.RecordingSlackPosterAdaptor
import dev.number6.slackposter.model.ChannelSummaryImageBuilder
import dev.number6.slackposter.model.PresentableChannelSummary
import org.junit.jupiter.api.Test

class SlackPosterComponentIntegrationTest {
    private var testee = DaggerTestSlackPosterComponent.create()

    @Test
    fun postCorrectChatToSlack() {
        val image: ChannelSummaryImage = ChannelSummaryImageBuilder.finalImage().build()
        val service = testee.handler()
        service.handleNewImage(image, FakeLambdaLogger())
        val http = testee.recordingHttpAdaptor() as RecordingSlackPosterAdaptor
        assertThat(http.posts).hasSize(1)
        val chat = http.posts[0]
        assertThat(chat.initialMessageLine).isEqualTo(PresentableChannelSummary(image).initialMessageLine)
        chat.attachments.all { a ->
            PresentableChannelSummary(image).attachments.any { a2 ->
                a2.pretext == a.pretext &&
                        a2.fields.contentEquals(a.fields)
            }
        }
    }

    internal class FakeLambdaLogger : LambdaLogger {
        override fun log(message: String) {
            println(message)
        }

        override fun log(message: ByteArray) {
            throw UnsupportedOperationException("doin' wut now?")
        }
    }
}